cql3/expr/expression.cc

```diff
@@ -1411,8 +1415,172 @@ constant evaluate(const expression& e, const query_options& options) {
     }, e);
 }
 
+// Checks whether a bound value of this type has to be reserialized
+static bool needs_to_be_reserialized(const abstract_type& type, const cql_serialization_format& sf) {
+    // If a value contains set or map, then this collection can be sent in the wrong order
```

Comment:
This can be in types.cc, the knowledge of serialization formats should be embedded there. 

Further, we can optimize it by computing it during type construction time and storing it in in a member.

Reply:
Do types ever change?
I think it's possible to add new field to a user type (?)

I was worried that this would be premature optimalization, the types aren't that big and there would be some new delicate bug prone logic introduced.

Reply:
It's possible to add new fields to user types, but they remain serialization compatible.

Serialization is not a property of the CQL front end, it's a property of the binary protocol between client and server and of the types themselves. It's really doesn't belong in expression.cc.

It's unlikely to ever change, since even if we add types, they won't exist in protocol v2.